src/Streamly/Internal/Data/Unfold/Resume/Type.hs

```diff
@@ -0,0 +1,168 @@
+-- |
+-- Module      : Streamly.Internal.Data.Unfold.Resume.Type
+-- Copyright   : (c) 2021 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Resumable unfolds.  See "Streamly.Internal.Data.Unfold.Resume" for
+-- introduction.
+--
+
+module Streamly.Internal.Data.Unfold.Resume.Type
+    (
+    -- * Type
+    Unfold (..)
+
+    -- * Unfolds
+    , nil
+    , nilM
+    , unfoldrM
+    , fromList
+
+    -- * Mapping
+    , lmap
+
+    -- * Nesting
+    , concat
+    )
+where
+
+#include "inline.hs"
+
+import Fusion.Plugin.Types (Fuse(..))
+import Streamly.Internal.Data.Stream.StreamD.Step (Step(..))
+import Prelude hiding (concat)
+
+------------------------------------------------------------------------------
+-- Type
+------------------------------------------------------------------------------
+
+-- | An @Unfold m a b@ is a generator of a stream of values of type @b@ from a
+-- seed of type 'a' in 'Monad' @m@.
+--
+-- /Internal/
+
+data Unfold m a b =
+    -- | @Unfold step inject extract@
+    forall s. Unfold (s -> m (Step s b)) (a -> m s) (s -> m (Maybe a))
+
+------------------------------------------------------------------------------
+-- Unfolds
+------------------------------------------------------------------------------
+
+-- XXX we do not need a Maybe, we can just use undefined or error. There is no
+-- state to extract in this case, and it should not be used.
```

Comment:
Is this is specific to `nilM`?

> we do not need a Maybe, we can just use undefined or error

What does this mean? We have a `Maybe` so might as well use it?